### PR TITLE
Getting rid of "module." heritage

### PR DESCRIPTION
**Pull request by BloodAxe**
# What this PR does

We do not store state_dict of the model with "module." prefix anymore

# Why this is important

- When training with DDP/DP mode, DistributedDataParallel/DataParallel wrappers adds this "module." prefix unintentionally to `state_dict`. So there goes discrepancy when the model is trained with or without DDP. Previously SG was solving this with a hack of adding artificial "module." prefix to all saved models in single GPU mode. However, this is a broken design

- Some of our checkpoints are saved with and some are saved without "module." which I believe the main cause we have this ugly NO_KEY_MATCHING checkpoint loading hack. The actual reasoning is lost in time, but I'm, pretty sure this is because of this discrepancy of saving models with and without DDP wrapper.

# Solution

1. This PR introduce `get_real_model` method to return the "real" model if it is wrapped with DP/DDP. Whenever there is a need to save state_dict one should use this method to unwrap the model and save state dict of the real one. 
2. Similarly, when loading a checkpoint same logic should be performed: get_real_model(net).load_state_dict(...)
4. Compatibility with existing checkpoints that may or may not has "module." prefix is secured buy checking whether all state_dict keys are starting with "module.". If so - we drop them and load checkpoint as usual. This handles the case when checkpoint was saved with DDP wrapper.

# Testing

Not properly tested yet
Components requiring special attention
- [x] Regular trainer
- [ ] KDTrainer
- [x] EMA
- [x] Pretrained models
- [x] Unit Tests
- [x] Integration Tests
- [x] Shortened Recipe Tests

# Risks of introducing breaking changes

No risk assessment has been done yet

# Related issues

https://github.com/Deci-AI/super-gradients/issues/1163
https://github.com/Deci-AI/super-gradients/issues/1153

### CODE REVIEW

**Review by Louis-Dupont — COMMENTED** [review]
Looks great, I am just not sure about the name `get_real_model`, I think it is pretty confusing. 
Maybe `unwrap_model`? It's also not that good, but maybe slightly more explicit? Any better idea?

**Comment by BloodAxe** [comment]
`unwrap_model` is good, I like this name)

**Review by Louis-Dupont — APPROVED** [review]
LGTM

File: Makefile
Comment (ofrimasad) [inline Makefile:1]: why was this file changed?

**Review by ofrimasad — APPROVED** [review]
LGTM 
not sure if the Makefile was changed by mistake

File: Makefile
Comment (BloodAxe) [reply Makefile:1]: The change of Makefile was due to wrong executable file. I copied these lines from circle CI, but  on other machines python 3.8 may not be available, so the right excutable is just `python`, which would guarantee we are running against default python from currently active environment.
I used this to run all unit/integration/recipe tests locally on tzgas

**Review by ofrimasad — DISMISSED** [review]
LGTM

**Review by Louis-Dupont — APPROVED** [review]
LGTM
